### PR TITLE
Introduce fun_run for running and logging Commands

### PR DESCRIPTION
**Pull request by schneems**
Explored in https://github.com/heroku/buildpacks-ruby/pull/120, these criteria were generated for running commands:

- A representation of the command being run should appear in logs, or the error, or both
- Stdout and Stderr from the command should appear in logs, or the error, but NOT both (if it runs).
- Make non-zero status an error
- Allow for debugging a command that was run after execution (i.e., get command program name, arguments, env, cwd and pass them to something like `WhichProblem`)
- Ideally only have to format `Command` once

That PR had 2 full implementations and an experiment. Ultimately I prefer the direction of the trait experiment and I chose to go in that direction: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=decd260cddc65598c2a1f8ad5d6ea422

The "name and yield" approach satisfied all my criteria by introducing only a single, purpose-built method. It takes a function that generates a name String, and then turns right back around and yields it back to the next function, which returns a Result with generic valid/invalid states.

This approach is not the most intuitive, but very flexible, and it satisfies all the criteria. This commit represents a full implementation of that trait-extension experiment.

<details>
  <summary>Archive of trait experiment</summary>

## Archive

Full contents of https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=decd260cddc65598c2a1f8ad5d6ea422 (for archival)

```
use std::process::{Command, Output};

/// Trait extension experiments
///
/// # Goals:
///
/// - A representation of the command being run should appear in logs, or the error, or both
/// - Stdout and Stderr from the command should appear in logs, or the error, but NOT both (if it runs).
/// - Make non-zero status an error
/// - Allow for debugging a command that was run after execution (i.e. get command program name, arguments, env, cwd and pass them to something like `WhichProblem`)
/// - Ideally only have to format `Command` once
///
/// Implemented two different ways with two different philosophies.
///
/// ## Punted on
///
/// - Stdout and Stderr from the command should appear in logs, or the error, but NOT both (if it runs).
///
/// This could be solved by forcing the user to wrap their error output in a enum as in prior experiments.
///
/// We could introduce some optional convienence trait extensions like `fn pretty_stream(&mut self)`
/// that does it for you automatically.
fn main() {
    println!("Hello, world!");

    // ## Peek and capture
    //
    // Two general purpose trait extensions, but requires manual name managment
    //
    // Here name must either be persisted as a shared mutable string or
    // by repetedly calling the same function (as shown below).
    //
    // Peek isn't really needed, but makes intent a tad bit cleaner
    // and allows us to pass a function directly to the other extension
    let _output = Command::new("bundle")
        .arg("install")
        .peek(|cmd| {
            println!("Running $ {}", display_cmd(cmd));
        })
        .command_output(Command::output)
        .map_err(|(cmd, error)| {
            println!("Pretend investigate cmd args {cmd:?}");
            ErrorPlaceholder::System(display_cmd(cmd), error)
        })
        .and_then(|(cmd, out)| {
            if out.status.success() {
                Ok(out)
            } else {
                println!("Pretend investigate cmd args {cmd:?}");
                Err(ErrorPlaceholder::NonZero(display_cmd(cmd), out))
            }
        })
        .unwrap();

    // ## Name and yield
    //
    // One very specific trait extension, requires extra indentation
    let _output = Command::new("bundle")
        .arg("install")
        .args(["--path", "lol"])
        .get_name_then_output(
            |cmd| format!("{cmd:?}"),
            |cmd, name| {
                println!("Running $ {name}");
                cmd.output()
                    .map_err(|error| {
                        println!("Pretend investigate cmd args {cmd:?}");
                        ErrorPlaceholder::System(name.clone(), error)
                    })
                    .and_then(|out| {
                        if out.status.success() {
                            Ok(out)
                        } else {
                            println!("Pretend investigate cmd args {cmd:?}");
                            Err(ErrorPlaceholder::NonZero(name.clone(), out))
                        }
                    })
            },
        );
}

fn display_cmd(cmd: &Command) -> String {
    cmd.get_program().to_string_lossy().to_string()
}

trait ComandOutputPlaceholder<F>
where
    F: FnOnce(&mut Command) -> Result<Output, std::io::Error>,
{
    fn command_output(
        &mut self,
        f: F,
    ) -> Result<(&mut Command, Output), (&mut Command, std::io::Error)>;
}

impl<F> ComandOutputPlaceholder<F> for Command
where
    F: FnOnce(&mut Command) -> Result<Output, std::io::Error>,
{
    fn command_output(
        &mut self,
        f: F,
    ) -> Result<(&mut Command, Output), (&mut Command, std::io::Error)> {
        match f(self) {
            Ok(output) => Ok((self, output)),
            Err(error) => Err((self, error)),
        }
    }
}

trait CommandPeekPlaceholder<F>
where
    F: FnOnce(&mut Command),
{
    fn peek(&mut self, f: F) -> &mut Self;
}

impl<F> CommandPeekPlaceholder<F> for Command
where
    F: FnOnce(&mut Command),
{
    fn peek(&mut self, f: F) -> &mut Self {
        f(self);
        self
    }
}

trait Lol<N, G, O, E>
where
    N: Fn(&mut Command) -> String,
    G: FnOnce(&mut Command, String) -> Result<O, E>,
{
    fn get_name_then_output(&mut self, get_name: N, get_output: G) -> Result<O, E>;
}

impl<N, G, O, E> Lol<N, G, O, E> for Command
where
    N: Fn(&mut Command) -> String,
    G: FnOnce(&mut Command, String) -> Result<O, E>,
{
    fn get_name_then_output(&mut self, get_name: N, get_output: G) -> Result<O, E> {
        let name = get_name(self);
        get_output(self, name)
    }
}

#[derive(Debug)]
enum ErrorPlaceholder {
    System(String, std::io::Error),
    NonZero(String, Output),
}
```
</details>

### CODE REVIEW

File: buildpacks/ruby/src/layers/bundle_download_layer.rs [hunk 1]
```diff
@@ -58,12 +62,21 @@ impl Layer for BundleDownloadLayer {
...
-            ],
-            &self.env,
-        )
-        .output()
-        .map_err(RubyBuildpackError::GemInstallBundlerCommandError)?;
+                &bin_dir.to_string_lossy(),
+            ])
+            .env_clear()
+            .envs(&self.env)
+            .name_and_then(fun_run::display, |name, cmd| {
```
[the remark this answers is not in the document]
Comment (joshwlewis) [reply buildpacks/ruby/src/layers/bundle_download_layer.rs:69]: Or maybe `.with_name().and_then(|name, cmd| { // do_stuff }`.

File: commons/src/fun_run.rs [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+///
+/// ```rust
+/// use std::process::Command;
+/// use commons::fun_run;
+///
+/// let name = fun_run::display(Command::new("bundle").arg("install"));
+/// assert_eq!(String::from("bundle install"), name);
+/// ```
+#[must_use]
+pub fn display(command: &mut Command) -> String {
```
Comment (joshwlewis) [inline commons/src/fun_run.rs:108]: Why not implement `Display` for `Command`? That way, you could `Command::new("bundle").arg("install").to_string()`.

**Review by joshwlewis — APPROVED** [review]
LGTM!

File: commons/src/fun_run.rs (hunk 2 above)
Comment (schneems) [reply commons/src/fun_run.rs:108]: We can't implement `Display` directly due to https://github.com/Ixrec/rust-orphan-rules. There's a way to do it that's hacky that involves creating our own trait that then implements that trait, but it's well...hacky.

I'm not opposed to that existing, I just also think people want to have the option of using a method if they want. 

The other concern is different people need different formatting. I want/need to have env vars in my displays.

**Review by Malax — APPROVED** [review]
As discussed in person, this is fine for the Ruby buildpack. If we would want to move this to a shared crate we can mull over the details and API if the time comes.

File: buildpacks/ruby/src/layers/bundle_download_layer.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (schneems) [reply buildpacks/ruby/src/layers/bundle_download_layer.rs]: I updated it to `cmd_map` and got rid of the `map_and_then` function. I think the API is better.

**Pull request merged**